joblib/parallel.py

```diff
@@ -585,8 +598,18 @@ def _dispatch(self, batch):
 
         dispatch_timestamp = time.time()
         cb = BatchCompletionCallBack(dispatch_timestamp, len(batch), self)
-        job = self._backend.apply_async(batch, callback=cb)
-        self._jobs.append(job)
+        with self._lock:
+            job_idx = len(self._jobs)
+            job = self._backend.apply_async(batch, callback=cb)
+            # If the callback is already done and we are not using a
+            # SequentialBackend, we need to reorder the jobs to get the correct
+            # order in the result. This can happen when a job is so quick
+            # that its callback is called before we get here.
+            if (not issubclass(type(self._backend), SequentialBackend) and
+                    cb.done):
+                self._jobs.insert(job_idx, job)
```

Comment:
Would `self._jobs.insert(job_idx, job)` not always work, i.e. you don't need the if clause?